connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaStatusBackingStore.java

```diff
@@ -0,0 +1,462 @@
+ *
+ * 3) It is (probably) safe to overwrite the state if there is no previous
+ *    value.
+ * 1) It is (probably) safe to overwrite the state if the previous value was
+ *    set by a worker with the same workerId.
+ * 2) It is (probably) safe to overwrite the previous state if the current
+ *    generation is higher than the previous .
+ *
+ * Basically all these conditions do is reduce the window for conflicts. They
+ * obviously cannot take into account in-flight requests.
+ *
+ */
+public class KafkaStatusBackingStore implements StatusBackingStore {
+    private static final Logger log = LoggerFactory.getLogger(KafkaStatusBackingStore.class);
+
+    public static final String STATUS_TOPIC_CONFIG = "status.storage.topic";
+
+    private static final String TASK_STATUS_PREFIX = "status-task-";
+    private static final String CONNECTOR_STATUS_PREFIX = "status-connector-";
+
+    public static final String STATE_KEY_NAME = "state";
+    public static final String MSG_KEY_NAME = "msg";
+    public static final String WORKER_ID_KEY_NAME = "worker_id";
+    public static final String GENERATION_KEY_NAME = "generation";
+
+    private static final Schema STATUS_SCHEMA_V0 = SchemaBuilder.struct()
+            .field(STATE_KEY_NAME, Schema.STRING_SCHEMA)
+            .field(MSG_KEY_NAME, SchemaBuilder.string().optional().build())
+            .field(WORKER_ID_KEY_NAME, Schema.STRING_SCHEMA)
+            .field(GENERATION_KEY_NAME, Schema.INT32_SCHEMA)
+            .build();
+
+    private final Time time;
+    private final Converter converter;
+    private final Table<String, Integer, CacheEntry<TaskStatus>> tasks;
+    private final Map<String, CacheEntry<ConnectorStatus>> connectors;
+
+    private String topic;
+    private KafkaBasedLog<String, byte[]> kafkaLog;
+    private int generation;
+
+    public KafkaStatusBackingStore(Time time, Converter converter) {
+        this.time = time;
+        this.converter = converter;
+        this.tasks = new Table<>();
+        this.connectors = new HashMap<>();
+    }
+
+    // visible for testing
+    KafkaStatusBackingStore(Time time, Converter converter, String topic, KafkaBasedLog<String, byte[]> kafkaLog) {
+        this(time, converter);
+        this.kafkaLog = kafkaLog;
+        this.topic = topic;
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        if (configs.get(STATUS_TOPIC_CONFIG) == null)
+            throw new ConnectException("Must specify topic for connector status.");
+        this.topic = (String) configs.get(STATUS_TOPIC_CONFIG);
+
+        Map<String, Object> producerProps = new HashMap<>();
+        producerProps.putAll(configs);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
+        producerProps.put(ProducerConfig.ACKS_CONFIG, "all");
```

Comment:
Hmm, I'm realizing this might be missing for other uses of KafkaBasedLog, but should we be setting configs to match MirrorMaker-like setings, i.e. infinite retries, 1 max in flight requests? Otherwise status update order can get mixed up.